File: exam.py
import os
import sys
import platform
import time

def clear_screen():
    """Clears the terminal screen."""
    if platform.system() == "Windows":
        os.system('cls')
    else:
        os.system('clear')

def load_exam(file_path):
    """Loads the exam questions from the specified file."""
    questions = []
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            for line in file:
                if line.strip():  # Skip empty lines
                    parts = line.strip().split('|')
                    if len(parts) == 3:
                        question, options, answer = parts
                        questions.append({
                            'question': question.strip(),
                            'options': options.split(';'),
                            'answer': answer.strip().lower()
                        })
    except FileNotFoundError:
        print(f"Error: {file_path} not found.")
    return questions

def list_exam_files(directory):
    """Lists all .exam files in the given directory."""
    return [f for f in os.listdir(directory) if f.endswith('.exam')]

def run_exam(questions):
    """Runs the exam and collects user answers."""
    score = 0
    incorrect_answers = []
    clear_screen()
    print("\n--- Starting the Exam ---")
    for idx, q in enumerate(questions):
        clear_screen()
        print(f"\nQuestion {idx + 1}/{len(questions)}: {q['question']}")
        for i, option in enumerate(q['options'], start=1):
            print(f"  {chr(96 + i)}. {option.strip()}")
        user_answer = input("Your answer (a/b/c/d/e): ").lower()
        if user_answer == q['answer']:
            print("Correct!")
            score += 1
        else:
            print(f"Wrong! The correct answer was: {q['answer']}")
            incorrect_answers.append((q['question'], user_answer, q['answer']))
        print(f"Current Score: {score}/{idx + 1}")
        input("\nPress any key to continue...")  # Pause before next question
    return score, incorrect_answers

def view_incorrect_answers(incorrect_answers):
    """Displays questions the user answered incorrectly."""
    clear_screen()
    if not incorrect_answers:
        print("\nNo incorrect answers. Well done!")
    else:
        print("\n--- Incorrect Answers ---")
        for idx, (question, user_answer, correct_answer) in enumerate(incorrect_answers, start=1):
            print(f"\n{idx}. {question}")
            print(f"   Your answer: {user_answer}")
            print(f"   Correct answer: {correct_answer}")
    input("\nPress any key to return to the menu...")

def main():
    while True:
        if len(sys.argv) > 1:  # If a file is dragged
            file_path = sys.argv[1]
            print(f"Using dragged file: {file_path}")
        else:
            # List all .exam files in the current directory
            exam_files = list_exam_files(os.getcwd())
            if not exam_files:
                print("No .exam files found in the current directory. Exiting...")
                return

            print("Available exams:")
            for idx, exam_file in enumerate(exam_files):
                print(f"{idx + 1}. {exam_file}")
            
            # Let the user select an exam
            choice = int(input("Choose an exam by number: ")) - 1
            if choice < 0 or choice >= len(exam_files):
                print("Invalid choice. Exiting...")
                return
            
            file_path = exam_files[choice]
            print(f"Using selected file: {file_path}")
        
        questions = load_exam(file_path)
        if not questions:
            print("Failed to load questions. Exiting...")
            return

        while True:
            clear_screen()
            print(f"Loaded {len(questions)} questions from {file_path}.")
            score, incorrect_answers = run_exam(questions)

            # Show score
            clear_screen()
            print("\n--- Exam Finished ---")
            print(f"Your final score: {score}/{len(questions)}")
            percentage = (score / len(questions)) * 100
            print(f"Your performance: {percentage:.2f}%")
            if percentage >= 90:
                print("Excellent job!")
            elif percentage >= 75:
                print("Good work!")
            elif percentage >= 50:
                print("Needs improvement.")
            else:
                print("Better luck next time!")

            # End options
            print("\nWhat would you like to do next?")
            print("1. View incorrect answers")
            print("2. Restart this exam")
            print("3. Pick another exam")
            print("4. Exit")
            choice = input("Enter your choice: ")

            if choice == '1':
                view_incorrect_answers(incorrect_answers)
            elif choice == '2':
                continue
            elif choice == '3':
                break
            elif choice == '4':
                print("Goodbye!")
                return
            else:
                print("Invalid choice. Please try again.")
                time.sleep(2)

if __name__ == "__main__":
    main()

File: exam.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <random>
#include <cctype>
#include <cstdlib>
#include <filesystem>
#include <tuple>
#include <iomanip>

#ifdef _WIN32
#include <windows.h>
#else
#include <unistd.h>
#endif

namespace fs = std::filesystem;

// Function to clear the console screen
void clearScreen() {
#ifdef _WIN32
    system("cls");
#else
    system("clear");
#endif
}

// Struct to represent a question
struct Question {
    std::string questionText;
    std::vector<std::string> options;
    char answer; // Stores 'a', 'b', 'c', etc.
};

// Trim utility function (removes leading and trailing spaces)
std::string trim(const std::string& str) {
    auto start = str.find_first_not_of(" \t");
    auto end = str.find_last_not_of(" \t");
    return (start == std::string::npos) ? "" : str.substr(start, end - start + 1);
}

// Function to load exam questions from a file
std::vector<Question> loadExam(const std::string& filePath) {
    std::vector<Question> questions;
    std::ifstream file(filePath);

    if (!file) {
        std::cerr << "Error: File not found - " << filePath << std::endl;
        return questions;
    }

    std::string line;
    while (std::getline(file, line)) {
        if (line.empty()) continue;

        std::istringstream stream(line);
        std::string questionPart, optionsPart, answerPart;

        if (std::getline(stream, questionPart, '|') &&
            std::getline(stream, optionsPart, '|') &&
            std::getline(stream, answerPart, '|')) {

            Question question;
            question.questionText = questionPart; // Preserve original spaces in the question

            // Parse options and preserve spaces
            std::istringstream optionsStream(optionsPart);
            std::string option;
            while (std::getline(optionsStream, option, ';')) {
                question.options.push_back(option); // Preserve original spaces in options
            }

            // Trim only the correct answer part
            answerPart = trim(answerPart);

            // Parse and map answer to option letter (e.g., 'a', 'b', 'c')
            if (!answerPart.empty() && std::isdigit(answerPart[0])) {
                int answerIndex = std::stoi(answerPart) - 1; // Convert 1-based index to 0-based
                if (answerIndex >= 0 && answerIndex < static_cast<int>(question.options.size())) {
                    question.answer = 'a' + answerIndex; // Map to 'a', 'b', etc.
                } else {
                    std::cerr << "Warning: Invalid answer index for question: " << question.questionText << "\n";
                    question.answer = ' '; // Invalid answer
                }
            } else if (!answerPart.empty() && std::isalpha(answerPart[0])) {
                question.answer = std::tolower(answerPart[0]); // Direct mapping
            } else {
                std::cerr << "Warning: Invalid answer format for question: " << question.questionText << "\n";
                question.answer = ' '; // Invalid answer
            }

            questions.push_back(question);
        }
    }

    return questions;
}

// Function to scramble the order of questions
void scrambleQuestions(std::vector<Question>& questions) {
    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(questions.begin(), questions.end(), g);
}

// Function to list all `.exam` files in the current directory
std::vector<std::string> listExamFiles(const std::string& directory) {
    std::vector<std::string> examFiles;
    for (const auto& entry : fs::directory_iterator(directory)) {
        if (entry.path().extension() == ".exam") {
            examFiles.push_back(entry.path().filename().string());
        }
    }
    return examFiles;
}

// Function to run the exam
std::pair<int, std::vector<std::tuple<std::string, char, char>>>
runExam(const std::vector<Question>& questions) {
    int score = 0;
    std::vector<std::tuple<std::string, char, char>> incorrectAnswers;

    clearScreen();
    std::cout << "\n--- Starting the Exam ---\n";

    for (size_t i = 0; i < questions.size(); ++i) {
        const auto& q = questions[i];
        clearScreen();
        std::cout << "\nQuestion " << (i + 1) << "/" << questions.size() << ": " << q.questionText << "\n";

        for (size_t j = 0; j < q.options.size(); ++j) {
            std::cout << "  " << char('a' + j) << ". " << q.options[j] << "\n";
        }

        char userAnswer;
        std::cout << "Your answer (a/b/c/d/e): ";
        std::cin >> userAnswer;
        userAnswer = std::tolower(userAnswer);

        if (userAnswer == q.answer) {
            std::cout << "Correct!\n";
            ++score;
        } else {
            std::cout << "Wrong! The correct answer was: " << q.answer << "\n";
            incorrectAnswers.emplace_back(q.questionText, userAnswer, q.answer);
        }
        std::cout << "Current Score: " << score << "/" << (i + 1) << "\n";
        std::cout << "\nPress Enter to continue...";
        std::cin.ignore();
        std::cin.get();
    }

    return { score, incorrectAnswers };
}

// Function to view incorrect answers
void viewIncorrectAnswers(const std::vector<std::tuple<std::string, char, char>>& incorrectAnswers) {
    clearScreen();
    if (incorrectAnswers.empty()) {
        std::cout << "\nNo incorrect answers. Well done!\n";
    } else {
        std::cout << "\n--- Incorrect Answers ---\n";
        for (size_t i = 0; i < incorrectAnswers.size(); ++i) {
            const auto& [question, userAnswer, correctAnswer] = incorrectAnswers[i];
            std::cout << "\n" << (i + 1) << ". " << question << "\n";
            std::cout << "   Your answer: " << userAnswer << "\n";
            std::cout << "   Correct answer: " << correctAnswer << "\n";
        }
    }
    std::cout << "\nPress Enter to return to the menu...";
    std::cin.ignore();
    std::cin.get();
}

int main() {
    while (true) {
        // List `.exam` files in the current directory
        std::vector<std::string> examFiles = listExamFiles(".");
        if (examFiles.empty()) {
            std::cout << "No .exam files found in the current directory. Exiting...\n";
            return 1;
        }

        std::cout << "Available exams:\n";
        for (size_t i = 0; i < examFiles.size(); ++i) {
            std::cout << (i + 1) << ". " << examFiles[i] << "\n";
        }

        std::cout << "Choose an exam by number: ";
        int choice;
        std::cin >> choice;
        if (choice < 1 || choice > static_cast<int>(examFiles.size())) {
            std::cout << "Invalid choice. Exiting...\n";
            return 1;
        }

        std::string selectedFile = examFiles[choice - 1];
        std::vector<Question> questions = loadExam(selectedFile);

        if (questions.empty()) {
            std::cout << "Failed to load questions. Exiting...\n";
            return 1;
        }

        // Ask if the user wants to scramble the questions
        std::cout << "Do you want to scramble the exam questions? (y/n): ";
        char scrambleChoice;
        std::cin >> scrambleChoice;
        if (std::tolower(scrambleChoice) == 'y') {
            scrambleQuestions(questions);
        }

        while (true) {
            clearScreen();
            std::cout << "Loaded " << questions.size() << " questions from " << selectedFile << ".\n";
            auto [score, incorrectAnswers] = runExam(questions);

            // Display final score
            clearScreen();
            std::cout << "\n--- Exam Finished ---\n";
            std::cout << "Your final score: " << score << "/" << questions.size() << "\n";
            double percentage = (static_cast<double>(score) / questions.size()) * 100.0;
            std::cout << "Your performance: " << percentage << "%\n";

            if (percentage >= 90) std::cout << "Excellent job!\n";
            else if (percentage >= 75) std::cout << "Good work!\n";
            else if (percentage >= 50) std::cout << "Needs improvement.\n";
            else std::cout << "Better luck next time!\n";

            // Menu options
            std::cout << "\nWhat would you like to do next?\n";
            std::cout << "1. View incorrect answers\n";
            std::cout << "2. Restart the same exam\n";
            std::cout << "3. Exit\n";

            int menuChoice;
            std::cin >> menuChoice;

            if (menuChoice == 1) {
                viewIncorrectAnswers(incorrectAnswers);
            } else if (menuChoice == 2) {
                break; // Restart the exam
            } else {
                return 0;
            }
        }
    }

    return 0;
}
